src/generator.py

```diff
@@ -0,0 +1,148 @@
+"""
+Main Class that abstracts Test Data Generation API.
+It represents first layer of contact with the user code
+and should handle:
+    - Tuning Global Variables
+    - Generation of Test Data using GA and PDG
+"""
+
+from population import Population
+from ipdg import IPDG
+from coveragetable import CoverageTable
+from sourcecode import SourceCode
+from testcase import TestCase
+from predicate import Predicate
+from constraint import Constraint
+
+import const
+from typing import List
+import random
+
+
+class Generator:
+    def validate_global_variables(self):
+        if self.seed_size < 2 or self.seed_size > 10:
+            raise AssertionError
+        if self.population_size < 1 or self.population_size > 500:
+            raise AssertionError
+        if self.rate_crossover < 0.0 or self.rate_crossover > 0.9:
+            raise AssertionError
+        if self.rate_mutation < 0.0 or self.rate_mutation > 0.9:
+            raise AssertionError
+        if self.max_iter < 1 or self.max_iter > 1000:
+            raise AssertionError
+        if self.wait_iter < 0 or self.wait_iter > 1000:
+            raise AssertionError
+        if self.dependency_mode < 0 or self.dependency_mode > 3:
+            raise AssertionError
+
+    def __init__(
+        self,
+        seed_size: int = const.DEF_SEED_SIZE,
+        population_size: int = const.DEF_POP_SIZE,
+        rate_crossover: float = const.DEF_RATE_CROSSOVER,
+        rate_mutation: float = const.DEF_RATE_MUTATION,
+        max_iter: int = const.DEF_MAX_ITER,
+        wait_iter: int = const.DEF_WAIT_ITER,
+        dependency_mode: int = const.DEF_DEPENDENCY_MODE,
+        test_type: int = const.DEF_TEST_TYPE,
+    ):
+        self.seed_size = seed_size
+        self.population_size = population_size
+        self.rate_crossover = rate_crossover
+        self.rate_mutation = rate_mutation
+        self.max_iter = max_iter
+        self.wait_iter = wait_iter
+        self.dependency_mode = dependency_mode
+        self.test_type = test_type
+        self.validate_global_variables()
+
+    def generate_tests(self, source_code: SourceCode):
+        ipdg = IPDG(source_code)
+        ct = CoverageTable(source_code)
+
+
+        # Random Seeding CT
+        random_seed = Population()
+        random_seed.initial_population(self.seed_size, self.test_type)
+        ct.update(random_seed)        
+
+        dependency_mode = 0
+
+        #for Statistics
+        generations_cnt = 0
+
+        while True:
+            target: Predicate = ct.get_target_branch()
```

Comment:
I added pdg parameter for get_target_branch() as it was really needed for that

Reply:
why do we need pdg for get_target_branch()? it should be solely tied to coveragetable isn't it?

Reply:
it depends on eases, which are calculated based on coverages and paths(of pdg)

Reply:
is it pdg or cfd?

Reply:
so ultimately init of CoverageTable should include both pdg and source_code right?

Reply:
no, only pdg i guess

Reply:
FORGET ABOUT THAT, I FIXED MY PART